src/platform/backends/shared/base_availability_zone_manager.cpp

```diff
@@ -0,0 +1,236 @@
+            zone_names.insert(entry.path().stem());
+        }
+        if (zone_names.empty())
+        {
+            mpl::log(mpl::Level::info, category, "no zones found, using defaults");
+            return default_zones;
+        }
+        return zone_names;
+    };
+
+    const auto zone_names = get_zone_names();
+
+    const auto file_type = MP_FILEOPS.status(file_path, err).type();
+    if (err && file_type != fs::file_type::not_found)
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "AZ manager file {:?} is not accessible: {}.",
+            file_path.string(),
+            err.message(),
+        };
+    }
+
+    if (file_type == fs::file_type::not_found)
+    {
+        mpl::log(mpl::Level::info,
+                 category,
+                 fmt::format("AZ manager file {:?} not found, using defaults", file_path.string()));
+        automatic_zone = *zone_names.begin();
+        serialize();
+        return;
+    }
+
+    if (file_type != fs::file_type::regular)
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "AZ manager file {:?} is not a regular file.",
+            file_path.string(),
+        };
+    }
+
+    mpl::log(mpl::Level::info, category, fmt::format("reading AZ manager from file {:?}", file_path.string()));
+    const auto file = MP_FILEOPS.open_read(file_path);
+    if (file->fail())
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "failed to open AZ manager file {:?} for reading: {}",
+            file_path.string(),
+            std::strerror(errno),
+        };
+    }
+
+    const std::string data{std::istreambuf_iterator{*file}, {}};
+    const auto qdata = QString::fromStdString(data).toUtf8();
+    const auto json = QJsonDocument::fromJson(qdata).object();
+
+    if (const auto json_automatic_zone = json[automatic_zone_key].toString().toStdString();
+        zone_names.find(json_automatic_zone) == zone_names.end())
+    {
+        mpl::log(mpl::Level::warning,
+                 category,
+                 fmt::format("automatic zone {:?} not known, using default", json_automatic_zone));
+        automatic_zone = *zone_names.begin();
+    }
+    else
+    {
+        automatic_zone = json_automatic_zone;
+    }
+
+    serialize();
+
+    for (const auto& name : zone_names)
+        zones[name] = std::make_unique<BaseAvailabilityZone>(name, zones_directory);
+}
+
+AvailabilityZone& BaseAvailabilityZoneManager::get_zone(const std::string& name)
+{
+    if (const auto it = zones.find(name); it != zones.end())
+        return *it->second;
+    throw AvailabilityZoneNotFound{name};
+}
+
+std::string BaseAvailabilityZoneManager::get_automatic_zone_name()
+{
+    const std::unique_lock lock{mutex};
+    // round-robin until we get an available zone
+    const auto start = zones.find(automatic_zone);
+    auto current = start;
+    do
+    {
+        if (current->second->is_available())
```

Comment:
It might be a good idea to do a structured binding here to give current->first && current->second proper names.